File: tennis-game/tennis-game-kotlin-arrow/src/main/kotlin/GamePlay.kt
import GameInteraction.readPlayer
import GameInteraction.showGameScore
import arrow.fx.IO
import arrow.fx.extensions.fx
import arrow.optics.Lens

fun playTennisGame(game: Game): IO<Game> =
    IO.fx {
        val playerScoring = !readPlayer()

        val updatedGame = trackScoredPoint(playerScoring, game)

        !showGameScore(updatedGame)

        when {
            gameCompleted(updatedGame) -> updatedGame
            else -> !playTennisGame(updatedGame)
        }
    }

val trackScoredPoint: (ScoringPlayer, Game) -> (Game) = { scoringPlayer, game ->
    when (scoringPlayer) {
        Player1 -> update(game, gameToPlayer1Score, gameToPlayer2Score)
        Player2 -> update(game, gameToPlayer2Score, gameToPlayer1Score)
    }
}

private val update: (
    game: Game,
    scoringPlayerScore: Lens<Game, Score>,
    opponentPlayerScore: Lens<Game, Score>
) -> Game = { game, scoringPlayerScore, opponentPlayerScore ->
    when (scoringPlayerScore.get(game)) {
        Love -> scoringPlayerScore.set(game, Fifteen)
        Fifteen -> scoringPlayerScore.set(game, Thirty)
        Thirty -> scoringPlayerScore.set(game, Forty)
        Forty -> when (opponentPlayerScore.get(game)) {
            Advantage -> opponentPlayerScore.set(game, Forty)
            Forty -> scoringPlayerScore.set(game, Advantage)
            else -> scoringPlayerScore.set(game, Wins)
        }
        Advantage -> scoringPlayerScore.set(game, Wins)
        Wins -> game
    }
}

val gameCompleted: (Game) -> Boolean = { game ->
    gameToPlayer1Score.get(game) == Wins || gameToPlayer2Score.get(game) == Wins
}

File: tennis-game/swift-bow-tennis-game/swift-bow-tennis-game/GamePlay.swift
//
//  GamePlay.swift
//  swift-bow-tennis-game
//
//  Created by Fabrizio Duroni on 08.09.20.
//  Copyright © 2020 Fabrizio Duroni. All rights reserved.
//

import Foundation
import BowOptics
import BowEffects

func trackScoredPoint(scoringPlayer: ScoringPLayer, game: Game) -> IO<Never, Game> {
    return IO.invoke { trackScorePointFor(scoringPlayer: scoringPlayer, game: game) }
}

func trackScorePointFor(scoringPlayer: ScoringPLayer, game: Game) -> Game {
    
    switch (scoringPlayer) {
    case .Player1: return update(game: game, scoringPlayerScore: gameToPlayer1Score, opponentPlayerScore: gameToPlayer2Score)
    case .Player2: return update(game: game, scoringPlayerScore: gameToPlayer2Score, opponentPlayerScore: gameToPlayer1Score)
    }
}

func completed(game: Game) -> Bool {
    return gameToPlayer1Score.get(game) == .Wins || gameToPlayer2Score.get(game) == .Wins
}

private func update(game: Game, scoringPlayerScore: Lens<Game, Score>, opponentPlayerScore: Lens<Game, Score>) -> Game {
    switch (scoringPlayerScore.get(game)) {
    case .Love: return scoringPlayerScore.set(game, .Fifteen)
    case .Fifteen: return scoringPlayerScore.set(game, .Thirty)
    case .Thirty: return scoringPlayerScore.set(game, .Forty)
    case .Forty:
        switch (opponentPlayerScore.get(game)) {
            case .Advantage: return opponentPlayerScore.set(game, .Forty)
            case .Forty: return scoringPlayerScore.set(game, .Advantage)
            default: return scoringPlayerScore.set(game, .Wins)
        }
    case .Advantage: return scoringPlayerScore.set(game, .Wins)
    case .Wins: return scoringPlayerScore.set(game, .Wins)
    }
}
